### PR TITLE
feat(card): added pictogram

### PR DESCRIPTION
**Pull request by RaphaelAmadeu**
### Related Ticket(s)

#4447

### Description

Added the abillity to use pictograms on `Card`.

Added a new story and knobs to test this functionality.

### CODE REVIEW

File: packages/web-components/src/components/card/card.ts [hunk 1]
```diff
@@ -140,6 +177,12 @@ class DDSCard extends StableSelectorMixin(DDSLink) {
   @property()
   href = '';
 
+  /**
+   * Pictogram position
+   */
+  @property({ attribute: 'pictogram-position', reflect: true })
+  pictogramPosition = 'top';
```
Comment (asudoh) [inline packages/web-components/src/components/card/card.ts:184]: Can we introduce an enum and use it here? A note is that we recently moved all enum definitions to `src/components/{{name-of-component}}/defs.ts`. Please refer to how `<dds-link-list>` defines its enum.

**Review by asudoh — COMMENTED** [review]
A nice change @RaphaelAmadeu!

**Comment by asudoh** [comment]
@RaphaelAmadeu Thank you for the update! It appears that the heading text is gone once I change the pictogram placement from bottom to top. Do you see the same...?

**Comment by jeffchew** [comment]
CTA Card in general looks busted:

Screen Shot 2020-12-21 at 1 37 10 PM

Is this change only for Web Components or we're also going to have for React too?

**Comment by asudoh** [comment]
I now see two things in the code. One is `slotchange` event for `<slot name="pictogram">` is fired on one that is stale. The ideal fix is avoiding re-rendering `<slot name="pictogram">` upon change in pictogram position, but doing so seems to make the codebase complex while the use case of dynamically changing the pictogram position is rare.

Another is that copy content is only rendered when pictogram is placed on bottom, that seems to connect to @jeffchew's finding. Is it intentional...?

A quick fix for the former would be:

```diff
diff --git a/packages/web-components/src/components/card/card.ts b/packages/web-components/src/components/card/card.ts
index 0a19a3a26..e56b0489a 100644
--- a/packages/web-components/src/components/card/card.ts
+++ b/packages/web-components/src/components/card/card.ts
@@ -75,12 +75,16 @@ class DDSCard extends StableSelectorMixin(DDSLink) {
    * Handles `slotchange` event.
    */
   protected _handleSlotChange({ target }: Event) {
-    const { name } = target as HTMLSlotElement;
+    const { pictogramPosition: currentPictogramPosition } = this;
+    const { dataset, name } = target as HTMLSlotElement;
+    const { pictogramPosition } = dataset;
+    if (!pictogramPosition || pictogramPosition === currentPictogramPosition) {
       const hasContent = (target as HTMLSlotElement)
         .assignedNodes()
         .some(node => node.nodeType !== Node.TEXT_NODE || node!.textContent!.trim());
       this[slotExistencePropertyNames[name] || '_hasCopy'] = hasContent;
     }
+  }
 
   /**
    * @returns The copy content.
@@ -132,7 +136,7 @@ class DDSCard extends StableSelectorMixin(DDSLink) {
           </p>
           ${this.pictogramPosition === PICTOGRAM_POSITION.TOP
             ? html`
-                <slot name="pictogram" @slotchange="${handleSlotChange}"></slot>
+                <slot name="pictogram" data-pictogram-position="${PICTOGRAM_POSITION.TOP}" @slotchange="${handleSlotChange}"></slot>
               `
             : ''}
           ${this.pictogramPosition !== PICTOGRAM_POSITION.TOP || !hasPictogram
@@ -148,7 +152,7 @@ class DDSCard extends StableSelectorMixin(DDSLink) {
           <slot name="footer"></slot>
           ${this.pictogramPosition === PICTOGRAM_POSITION.BOTTOM
             ? html`
-                <slot name="pictogram" @slotchange="${handleSlotChange}"></slot>
+                <slot name="pictogram" data-pictogram-position="${PICTOGRAM_POSITION.BOTTOM}" @slotchange="${handleSlotChange}"></slot>
               `
             : ''}
           ${hasPictogram && this.pictogramPosition === PICTOGRAM_POSITION.TOP
```

File: packages/web-components/src/components/card/defs.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/**
+ * @license
+ *
+ * Copyright IBM Corp. 2020
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+export enum PICTOGRAM_POSITION {
```
Comment (asudoh) [inline packages/web-components/src/components/card/defs.ts:10]: To align to `ICON_PLACEMENT` for `<dds-card-footer>`:

```suggestion
export enum PICTOGRAM_PLACEMENT {
```

File: packages/web-components/src/components/card/defs.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+   * Top pictogram
+   */
+  TOP = 'top',
+  /**
+   * Bottom pictogram
+   */
+  BOTTOM = 'bottom',
+}
+
+export default PICTOGRAM_POSITION;
```
Comment (asudoh) [inline packages/web-components/src/components/card/defs.ts:21]: To align to `ICON_PLACEMENT` for `<dds-card-footer>`:

```suggestion
export default PICTOGRAM_PLACEMENT;
```

File: packages/web-components/src/components/card/card.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ import StableSelectorMixin from '../../globals/mixins/stable-selector';
 import DDSLink from '../link/link';
 import DDSCardFooter from './card-footer';
 import styles from './card.scss';
+import { PICTOGRAM_POSITION } from './defs';
```
Comment (asudoh) [inline packages/web-components/src/components/card/card.ts:18]: To align to `ICON_PLACEMENT` for `<dds-card-footer>`:

```suggestion
import { PICTOGRAM_PLACEMENT } from './defs';
```

File: packages/web-components/src/components/card/__stories__/card.stories.ts [hunk 5]
```diff
@@ -16,6 +16,7 @@ import ifNonNull from 'carbon-web-components/es/globals/directives/if-non-null';
 import readme from './README.stories.mdx';
 import textNullable from '../../../../.storybook/knob-text-nullable';
 import '../card';
+import { PICTOGRAM_POSITION } from '../defs';
```
Comment (asudoh) [inline packages/web-components/src/components/card/__stories__/card.stories.ts:19]: To align to `ICON_PLACEMENT` for `<dds-card-footer>`:

```suggestion
import { PICTOGRAM_PLACEMENT } from '../defs';
```

File: packages/web-components/src/components/card/card.ts [hunk 6]
```diff
@@ -140,6 +190,12 @@ class DDSCard extends StableSelectorMixin(DDSLink) {
   @property()
   href = '';
 
+  /**
+   * Pictogram position
+   */
+  @property({ attribute: 'pictogram-position', reflect: true })
+  pictogramPosition = PICTOGRAM_POSITION.TOP;
```
Comment (asudoh) [inline packages/web-components/src/components/card/card.ts:197]: To align to `iconPlacement` for `<dds-card-footer>`:

```suggestion
  pictogramPlacement = PICTOGRAM_POSITION.TOP;
```

File: packages/web-components/src/components/card/__stories__/card.stories.ts [hunk 7]
```diff
@@ -45,6 +46,53 @@ const iconPlacement = {
...
+  [PICTOGRAM_POSITION.TOP]: PICTOGRAM_POSITION.TOP,
+  [PICTOGRAM_POSITION.BOTTOM]: PICTOGRAM_POSITION.BOTTOM,
+};
+
+export const Pictogram = ({ parameters }) => {
+  const { href, heading, copy, pictogramPlacement, eyebrow, inverse } = parameters?.props?.PictogramCard ?? {};
+  return html`
+    <dds-card
+      color-scheme=${inverse ? 'inverse' : ''}
+      pictogram-position="${pictogramPlacement}"
```
Comment (asudoh) [inline packages/web-components/src/components/card/__stories__/card.stories.ts:59]: To align to `ICON_PLACEMENT` for `<dds-card-footer>`:

```suggestion
      pictogram-placement="${pictogramPlacement}"
```

File: packages/web-components/src/components/card/card.scss [hunk 8]
```diff
@@ -18,3 +18,47 @@
     color: $inverse-link;
   }
 }
+
+:host(#{$dds-prefix}-card)[pictogram-position='bottom']:hover {
```
Comment (asudoh) [inline packages/web-components/src/components/card/card.scss:22]:
```suggestion
:host(#{$dds-prefix}-card)[pictogram-placement='bottom']:hover {
```

File: packages/web-components/src/components/card/card.scss [hunk 9]
```diff
@@ -18,3 +18,47 @@
...
+
+    .#{$prefix}--card__heading {
+      margin-bottom: 0;
+      flex: 1;
+      display: flex;
+      align-items: flex-end;
+    }
+  }
+
+  &[pictogram-position='bottom'] {
```
Comment (asudoh) [inline packages/web-components/src/components/card/card.scss:45]:
```suggestion
  &[pictogram-placement='bottom'] {
```

File: packages/web-components/src/components/card/card.ts [hunk 10]
```diff
@@ -63,15 +65,25 @@ class DDSCard extends StableSelectorMixin(DDSLink) {
...
   protected _handleSlotChange({ target }: Event) {
-    const { name } = target as HTMLSlotElement;
-    const hasContent = (target as HTMLSlotElement)
-      .assignedNodes()
-      .some(node => node.nodeType !== Node.TEXT_NODE || node!.textContent!.trim());
-    this[slotExistencePropertyNames[name] || '_hasCopy'] = hasContent;
+    const { pictogramPosition: currentPictogramPosition } = this;
+    const { dataset, name } = target as HTMLSlotElement;
+    const { pictogramPosition } = dataset;
+    if (!pictogramPosition || pictogramPosition === currentPictogramPosition) {
```
Comment (asudoh) [inline packages/web-components/src/components/card/card.ts:81]:
```suggestion
    const { pictogramPlacement: currentPictogramPlacement } = this;
    const { dataset, name } = target as HTMLSlotElement;
    const { pictogramPlacement } = dataset;
    if (!pictogramPlacement || pictogramPlacement === currentPictogramPlacement) {
```

File: packages/web-components/src/components/card/card.ts [hunk 11]
```diff
@@ -109,18 +121,56 @@ class DDSCard extends StableSelectorMixin(DDSLink) {
...
           </p>
-          <h3 ?hidden="${!hasHeading}" class="${prefix}--card__heading">
-            <slot name="heading" @slotchange="${handleSlotChange}"></slot>
-          </h3>
-          ${this._renderCopy()}
+          ${this.pictogramPosition === PICTOGRAM_POSITION.TOP
+            ? html`
+                <slot
+                  name="pictogram"
+                  data-pictogram-position="${PICTOGRAM_POSITION.TOP}"
```
Comment (asudoh) [inline packages/web-components/src/components/card/card.ts:141]:
```suggestion
                  data-pictogram-placement="${PICTOGRAM_PLACEMENT.TOP}"
```

File: packages/web-components/src/components/card/card.ts [hunk 12]
```diff
@@ -109,18 +121,56 @@ class DDSCard extends StableSelectorMixin(DDSLink) {
...
+                  </span>
+                </h3>
+              `
+            : null}
+          ${this.pictogramPosition === PICTOGRAM_POSITION.BOTTOM || !hasPictogram ? this._renderCopy() : ''}
+          ${this.pictogramPosition === PICTOGRAM_POSITION.BOTTOM
+            ? html`
+                <slot
+                  name="pictogram"
+                  data-pictogram-position="${PICTOGRAM_POSITION.BOTTOM}"
```
Comment (asudoh) [inline packages/web-components/src/components/card/card.ts:160]:
```suggestion
                  data-pictogram-placement="${PICTOGRAM_PLACEMENT.BOTTOM}"
```

**Review by asudoh — APPROVED** [review]
Thank you for the fix @RaphaelAmadeu! My apologies that we'd need a naming nit to align to other components. (Note that there are more lines to change - need search in the code to find the occurrences) Please let me know if you feel not in this PR and we can track the effort of change under #4100.

**Pull request merged**